### PR TITLE
Floc Userid System: add new ID module

### PR DESCRIPTION
**Pull request by SKOCHERI**
## Type of change

- [ ] Bugfix
- [ ] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

- test parameters for validating bids
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer
- [ ] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

### CODE REVIEW

**Comment by SKOCHERI** [comment]
https://github.com/prebid/Prebid.js/issues/6510

**Comment by patmmccann** [comment]
Why store the Floc, can't you just always get it from the browser?

**Review by patmmccann — COMMENTED** [review]
I think we'll want this to work when tokens are no longer necessary, if a missing token shouldn't throw an error 12 months from now, let's have it not throw one now.

**Comment by SKOCHERI** [comment]
> I think we'll want this to work when tokens are no longer necessary, if a missing token shouldn't throw an error 12 months from now, let's have it not throw one now.

@patmmccann, Can you elaborate on this. "token" is marked as a mandatory parameters.

**Comment by patmmccann** [comment]
The token only makes sense while floc is in trial
https://developer.chrome.com/origintrials/#/trials/active . As soon as this
is over, publishers should be able to use the module without a token.
Hence, token must be non-mandatory.

**Comment by patmmccann** [comment]
The trial is over in three months https://developer.chrome.com/origintrials/#/view_trial/213920982300098561

**Comment by jdwieland8282** [comment]
Is it obvious to anyone how the token actually gets used?  I just registered Sharedid.org for the origin trial, but based on this https://wicg.github.io/floc/#string-representation-of-the-interest-cohort-id there is no mention of how it actually should be used.

**Comment by patmmccann** [comment]
@jdwieland8282 see second paragraph 

https://github.com/GoogleChrome/OriginTrials/blob/gh-pages/developer-guide.md

**Comment by Viewtiful** [comment]
I agree with @patmmccann, it would be better from the start to entirely not rely on cookies

File: modules/userId/userId.md
Comment (patmmccann) [inline modules/userId/userId.md:188]: dont store the floc, get it fresh every time

Comment (smenzer) [reply modules/userId/userId.md:188]: if this is the approach, then all examples should remove the `storage` config object. even better, a warning/error should be thrown in the `getId` function if `storage` was set at all, forcing publishers NOT to set it...but that depends on how important it is to fetch a fresh value every time or not

File: modules/flocIdSystem.js
Comment (patmmccann) [inline modules/flocIdSystem.js:60]: As @smenzer points out on #6510 

The floc is not the token,  the Floc needs to come from a browser api call

File: modules/userId/userId.md
Comment (patmmccann) [reply modules/userId/userId.md:188]: Excellent point

Comment (patmmccann) [reply modules/userId/userId.md:188]: @icflournoy you may want to do the same thing,  as storage interferes with your functionality and RDM sets it

**Comment by pm-harshad-mane** [comment]
Hello @SKOCHERI , 
can you please add support to pass the version as well? Example below.
```
"eids": [
      {
        "source": "chrome.com",
        "uids": [
          {
            "id": "777454319066618",
            "atype": 1,
            "ext": {
              "ver": "chrome.1.1"
            }
          }
        ]
      }
    ]
```

File: modules/userId/userId.md
Comment (SKOCHERI) [reply modules/userId/userId.md]: @patmmccann @smenzer Have updated to get cohort id from browser and removed the storage.

File: test/spec/modules/eids_spec.js
Comment (patmmccann) [inline test/spec/modules/eids_spec.js:352]: chrome.com and web.dev have been proposed in #6510 
google.com seems like a candidate as well. 
does anyone have a preference?

File: modules/flocIdSystem.js
Comment (patmmccann) [inline modules/flocIdSystem.js:76]: this module should work with no token, as the token will no longer be required after July. It seems like this function should be behind the conditional for is the token defined.

File: integrationExamples/gpt/userId_example.html
Comment (patmmccann) [inline integrationExamples/gpt/userId_example.html:290]: I can't find the value of this token in your PR yet you say it is the default.

File: test/spec/modules/eids_spec.js
Comment (smenzer) [reply test/spec/modules/eids_spec.js:352]: i would vote for `chrome.com` as it's the browser providing this, and probably shouldn't be as broad as google.com. chrome.com vs web.dev, though, I don't have a strong opinion on

Comment (ian-lr) [reply test/spec/modules/eids_spec.js:352]: I'll second `chrome.com`.

File: integrationExamples/gpt/userId_example.html
Comment (SKOCHERI) [reply integrationExamples/gpt/userId_example.html:290]: @patmmccann For token, should the publisher register one based on their domain? For testing this module, I used the chrome version 91.0.4469.4 (Official Build) dev (x86_64) and enabled --enable-blink-features=InterestCohortAPI 
--enable-features="FederatedLearningOfCohorts:update_interval/10s/minimum_history_domain_size_required/1,FlocIdSortingLshBasedComputation,InterestCohortFeaturePolicy"

With these flags enabled, the token is not required and also for checking if "interest-cohort" is available, the chrome flag "Experimental Web Platform features" had to be enabled.

File: modules/flocIdSystem.js
Comment (SKOCHERI) [inline modules/flocIdSystem.js:79]: @smenzer @patmmccann For checking the chrome, should window.chrome check suffice? I was unable to test window.chrome.webstore || !!window.chrome.runtime as webstore/runtime is not available

File: modules/flocIdSystem.js
Comment (smenzer) [inline modules/flocIdSystem.js:11]: should we make sure the module name matches the file name?

File: modules/flocIdSystem.js
Comment (smenzer) [reply modules/flocIdSystem.js:79]: i'm not sure...sorry!

File: modules/flocIdSystem.md
Comment (smenzer) [inline modules/flocIdSystem.md:30]: this isn't consistent with the other examples or the module name; although I suggest changing the others to use `flocId` as the module name

File: modules/flocIdSystem.md
Comment (smenzer) [inline modules/flocIdSystem.md:33]: should you maybe add a check in `getId()` to look for storage being set, and if so, throw an error and fail out?

File: test/spec/modules/userId_spec.js
Comment (smenzer) [inline test/spec/modules/userId_spec.js:1463]: is this really a useful test? the module doesn't read from cookies and doesn't allow setting storage param (which the `getConfigMock()` is doing). I think you need to mock the `document.interestCohort()` instead

File: integrationExamples/gpt/userId_example.html
Comment (smenzer) [reply integrationExamples/gpt/userId_example.html:290]: I was also a bit confused like @patmmccann - in all the examples it says to use a Share ID token but doesn't mention what it is. If you don't need a token, then I think you should include a description for how to operate without one so it's clear for publishers.

I would suggest either: 
1. change the text in the examples to be something like `"INSERT_PUBLISHER_TOKEN"` and then provide links or a description of how to obtain it; OR
2. actually provide a valid token to use as a default

**Review by smenzer — CHANGES_REQUESTED** [review]
Some changes requested, but this is getting there!

File: integrationExamples/gpt/userId_example.html
Comment (ian-lr) [reply integrationExamples/gpt/userId_example.html:290]: @SKOCHERI @smenzer My understanding is that for the origin trial, browsers that are not explicitly setting flags (~5% of non-EEA Chrome instances) will need a token set before they will respond with a cohort ID. Setting flags overrides this behavior. 

To avoid publishers _having_ to enroll in the Origin trial process, I'd recommend setting a default token. I just obtained `A3dHTSoNUMjjERBLlrvJSelNnwWUCwVQhZ5tNQ+sll7y+LkPPVZXtB77u2y7CweRIxiYaGwGXNlW1/dFp8VMEgIAAAB+eyJvcmlnaW4iOiJodHRwczovL3NoYXJlZGlkLm9yZzo0NDMiLCJmZWF0dXJlIjoiSW50ZXJlc3RDb2hvcnRBUEkiLCJleHBpcnkiOjE2MjYyMjA3OTksImlzU3ViZG9tYWluIjp0cnVlLCJpc1RoaXJkUGFydHkiOnRydWV9` for sharedid.org origin. Can we set that as the fallback value?

File: modules/flocIdSystem.md
Comment (SKOCHERI) [reply modules/flocIdSystem.md:30]: Updated the module to "flocId"

File: modules/flocIdSystem.md
Comment (SKOCHERI) [reply modules/flocIdSystem.md:34]: Logging error if storage is configured

**Comment by smenzer** [comment]
after discussion in the Prebid Privacy & Identity PMC, it was agreed that a user id module could be used to capture the floc id, but that it should not be added to the eids array which is more difficult to parse out. so, we'll remove the `eids.js` and `eids.md` code from this PR.

since some ssp's may still want to pass the floc id in eids to themself, we should add a utility method to create an eids array and add the floc id, called `getUserIdsAsEidsWithFloc()`.

cc: @SKOCHERI @jdwieland8282 @ian-lr @pm-harshad-mane

**Comment by SKOCHERI** [comment]
> after discussion in the Prebid Privacy & Identity PMC, it was agreed that a user id module could be used to capture the floc id, but that it should not be added to the eids array which is more difficult to parse out. so, we'll remove the `eids.js` and `eids.md` code from this PR.
> 
> since some ssp's may still want to pass the floc id in eids to themself, we should add a utility method to create an eids array and add the floc id, called `getUserIdsAsEidsWithFloc()`.
> 
> cc: @SKOCHERI @jdwieland8282 @ian-lr @pm-harshad-mane

I will go ahead to remove the eid.js and eids.md and related test.

Do we want to add `getUserIdsAsEidsWithFloc()` as part of this PR?

**Comment by smenzer** [comment]
> Do we want to add `getUserIdsAsEidsWithFloc()` as part of this PR?

That would be ideal but not a requirement I would say.

File: integrationExamples/gpt/userId_example.html
Comment (smenzer) [inline integrationExamples/gpt/userId_example.html:291]: @SKOCHERI this is perfect, thanks!

File: modules/flocIdSystem.md
Comment (smenzer) [reply modules/flocIdSystem.md:34]: thanks!

File: modules/userId/userId.md
Comment (smenzer) [inline modules/userId/userId.md:119]: i think you copied this from the table above, so you have an extra copy of the token

```suggestion
                    token: "Registered token or use the default sharedid.org token" // Default sharedid.org token: "A3dHTSoNUMjjERBLlrvJSelNnwWUCwVQhZ5tNQ+sll7y+LkPPVZXtB77u2y7CweRIxiYaGwGXNlW1/dFp8VMEgIAAAB+eyJvcmlnaW4iOiJodHRwczovL3NoYXJlZGlkLm9yZzo0NDMiLCJmZWF0dXJlIjoiSW50ZXJlc3RDb2hvcnRBUEkiLCJleHBpcnkiOjE2MjYyMjA3OTksImlzU3ViZG9tYWluIjp0cnVlLCJpc1RoaXJkUGFydHkiOnRydWV9"
```

File: modules/userId/userId.md
Comment (smenzer) [inline modules/userId/userId.md:233]: same thing here

```suggestion
               token: "Registered token or use the default sharedid.org token" // Default sharedid.org token: "A3dHTSoNUMjjERBLlrvJSelNnwWUCwVQhZ5tNQ+sll7y+LkPPVZXtB77u2y7CweRIxiYaGwGXNlW1/dFp8VMEgIAAAB+eyJvcmlnaW4iOiJodHRwczovL3NoYXJlZGlkLm9yZzo0NDMiLCJmZWF0dXJlIjoiSW50ZXJlc3RDb2hvcnRBUEkiLCJleHBpcnkiOjE2MjYyMjA3OTksImlzU3ViZG9tYWluIjp0cnVlLCJpc1RoaXJkUGFydHkiOnRydWV9"
```

**Review by smenzer — APPROVED** [review]
Just a minor doc fix and then this is good to go

**Comment by aleksatr** [comment]
@smenzer I'm not sure what are the test-coverage requirements for id submodules, but this one is only at 8%?

**Comment by smenzer** [comment]
> @smenzer I'm not sure what are the test-coverage requirements for id submodules, but this one is only at 8%?

most of hte testing of he user id modules is in the userid_spec.js, so i think most modules are pretty well covered. 8% seems really low, though, so we should make sure the necessary tests are added here to get that up.

**Comment by patmmccann** [comment]
I don't like the planned obsolescence of this module to july 2021 when the origin trial ends. It needs to work without a token configured, not throw a warning without one,  and not try to insert the sharedid token unless configuration specifies the origin trial is still occurring

**Comment by jdwieland8282** [comment]
Hi @patmmccann that's not how I would characterize our approach.  We definitely see this module working post 7/21. It seems difficult to give the module behavior with a TTL. Like do this until ... then do that. Is it ok if we use the sharedid token by default (configurable of course) and throw a warning if omitted until July, then update the module in July?

**Comment by patmmccann** [comment]
That's not a good approach, just fix in July, as publishers have vary long and variable upgrade schedules. You guys can see that in your internal data I'm sure. Why not just not throw the warning in the code and put the warning in the doc that if you don't have a token, this isn't going to work from April 2021 until End of July 2021?

**Comment by jdwieland8282** [comment]
That's fine with me @patmmccann we will go with that approach.  Any other changes you'd like to see?  Are you ok using sharedid's token by default?

**Comment by patmmccann** [comment]
that's my only remaining concern, thanks!

**Review by patmmccann — APPROVED** [review]

File: integrationExamples/gpt/userId_example.html
Comment (aleksatr) [inline integrationExamples/gpt/userId_example.html:5]: @SKOCHERI can you revert the formatting changes you made to this file, so it includes only the floc-related changes?

File: modules/flocIdSystem.js
Comment (aleksatr) [inline modules/flocIdSystem.js:19]: maybe you want to end lines with `;` in this function as well, just to be consistent with the rest of the file.

File: modules/flocIdSystem.js
Comment (aleksatr) [inline modules/flocIdSystem.js:90]: @SKOCHERI did you maybe mean `configParams && (typeof configParams.token === 'string'` ? (judging by the change history, to get the same behavior)

**Comment by jdwieland8282** [comment]
@aleksatr Can you please review and approve if you are satisfied, we'd like to include this in todays pb.js release

File: modules/flocIdSystem.js
Comment (aleksatr) [reply modules/flocIdSystem.js]: I meant all the statements in this function, not just on this line :)

File: modules/flocIdSystem.js
Comment (smenzer) [inline modules/flocIdSystem.js:86]: out of curiosity, what's the reason to check for chrome separately from floc, considering it's `&&` together next and they are never used separately? couldn't it just be one long statement for `isFlocEnabled`?

**Review by aleksatr — APPROVED** [review]

**Comment by patmmccann** [comment]
@jdwieland8282 if we resolve conflicts i think this can ship before the release today

**Comment by SKOCHERI** [comment]
> @jdwieland8282 if we resolve conflicts i think this can ship before the release today

Will do it right away

**Comment by SKOCHERI** [comment]
@patmmccann  Have resolved the conflicts but the tests are failing. I don't see any user id tests failing. Have tried rerunning multiple times. Do you have any pointers?

**Comment by patmmccann** [comment]
Those tests are unrelated,  also failing on master

**Comment by paulandrewfarrow** [comment]
Any idea when this will start showing up at https://docs.prebid.org/download.html?

**Comment by ian-lr** [comment]
Hey @paulandrewfarrow I think this needed a change to the download page; I've just made a PR here which should address that once merged: https://github.com/prebid/prebid.github.io/pull/3005

**Comment by jdwieland8282** [comment]
docs PR https://github.com/prebid/prebid.github.io/pull/3006

**Pull request merged**
